README.md
## driver-ipfs (uplink v1.0.5)

[![Codacy Badge](https://api.codacy.com/project/badge/Grade/f855c19791c240b6b7b930fa712bd9ed)](https://app.codacy.com/gh/storj-thirdparty/driver-ipfs?utm_source=github.com&utm_medium=referral&utm_content=storj-thirdparty/driver-ipfs&utm_campaign=Badge_Grade_Dashboard)
[![Go Report Card](https://goreportcard.com/badge/github.com/storj-thirdparty/driver-ipfs)](https://goreportcard.com/report/github.com/storj-thirdparty/driver-ipfs)
![Cloud Build](https://storage.googleapis.com/storj-utropic-services-badges/builds/driver-ipfs/branches/master.svg)


## Overview

The IPFS Driver connects to an IPFS server, takes a backup of the specified files and uploads the backup data on Storj network.

```bash
Usage:
  driver-ipfs [command] <flags>

Available Commands:
  help        Help about any command
  store       Command to upload data to a Storj V3 network.
  download    Command to download data from a Storj V3 network.
  version     Prints the version of the tool

```

`store` - Connect to the specified IPFS (default: `ipfs_property.json`). Back-up of the IPFS is generated using tooling provided by IPFS and then uploaded to the Storj network.  Connect to a Storj v3 network using the access specified in the Storj configuration file (default: `storj_config.json`).

`download` - Connect to the specified IPFS (default: `ipfs_property.json`). Connect to a Storj v3 network using the access specified in the Storj configuration file (default: `storj_config.json`). Download the data using file hash and download location specified in the Storj download configuration file (default: `storj_download.json`).

Sample configuration files are provided in the `./config` folder.

##### Note: This driver can only be used for uploading and downloading files, directories are not yet supported.

## Requirements and Install

To build from scratch, [install the latest Go](https://golang.org/doc/install#install).

> Note: Ensure go modules are enabled (GO111MODULE=on)



#### Option #1: clone this repo (most common)

To clone the repo

```
git clone https://github.com/storj-thirdparty/driver-ipfs.git
```

Then, build the project using the following:

```
cd driver-ipfs
go build
```

#### Option #2:  ``go get`` into your gopath

 To download the project inside your GOPATH use the following command:

```
go get github.com/storj-thirdparty/driver-ipfs
```

## Connect to IPFS Server

Make sure you are connected to IPFS server. If not, run the ipfs daemon in another `terminal` to join your node to the public network:

```
$ ipfs daemon
```


## Run (short version)

Once you have built the project run the following commands as per your requirement:

##### Get help

```
$ ./driver-ipfs --help
```

##### Check version

```
$ ./driver-ipfs --version
```

##### Upload data from local storage using ipfs and upload to Storj

```
$ ./driver-ipfs store
```

##### Download backup from Storj using earlier created ipfs file hash

```
$ ./driver-ipfs download
```

## Documentation

For more information on runtime flags, configuration, testing, and diagrams, check out the [Detail](//github.com/storj-thirdparty/driver-ipfs/wiki/Home) or jump to:

* [Config Files](//github.com/storj-thirdparty/driver-ipfs/wiki/#config-files)
* [Run (long version)](//github.com/storj-thirdparty/driver-ipfs/wiki/#run)
* [Testing](//github.com/storj-thirdparty/driver-ipfs/wiki/#testing)
* [Data-Flow Diagram (Upload)](//github.com/storj-thirdparty/driver-ipfs/wiki/#data-flow-diagram-upload)
* [Data-Flow Diagram (Download)](//github.com/storj-thirdparty/driver-ipfs/wiki/#data-flow-diagram-download)
* [Work-Flow Diagram (Upload)](//github.com/storj-thirdparty/driver-ipfs/wiki/#work-flow-diagram-upload)
* [Work-Flow Diagram (Download)](//github.com/storj-thirdparty/driver-ipfs/wiki/#work-flow-diagram-download)


CHANGELOG.md
# connector-nextcloud Changelog

## [1.0.5] - 18-08-2020
### Changelog:
* Resolved upload path issue.

## [1.0.5] - 27-05-2020
### Changelog:
* Added cobra cli for user interface.
* Restructured project based on the requirements for cobra cli.
* Changed arguments to optional flags.
* Added `--accesskey` flag and removed `key`, `test`, `parse` and `debug` flags.
* Added `--storj` flag to set storj config file path and `--ipfs` to set ipfs config file path.
* Fixed bugs

## [1.0.0] - 12-02-2020

test/doc-check.sh
## verify documentation is present. 
## cd is /workspace
HOME_MD="./README.assets/HOME.md"
README_MD="./README.md"
if [ -f $HOME_MD ] && [ -f $README_MD ]; then
    echo "doco check PASSED"
    exit 0
fi
echo "doco check FAILED.  Missing either" $README_MD "or" $HOME_MD
exit 1


cmd/ipfs.go
package cmd

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"os"
	"path/filepath"
	"strconv"

	shell "github.com/ipfs/go-ipfs-api"
)

// ConfigIpfs defines the variables and types.
type ConfigIpfs struct {
	HostName  string `json:"hostName"`
	Port      string `json:"port"`
	Path      string `json:"path"`
	ChunkSize string `json:"chunkSize"`
}

// LoadIpfsProperty reads and parses the JSON file
// that contain a IPFS instance's property.
// and returns all the properties as an object.
func LoadIpfsProperty(fullFileName string) ConfigIpfs {
	var configIpfs ConfigIpfs

	// Open and read the file
	fileHandle, err := os.Open(filepath.Clean(fullFileName))
	if err != nil {
		log.Fatal("Can't open the file : ", err)
	}

	jsonParser := json.NewDecoder(fileHandle)
	if err = jsonParser.Decode(&configIpfs); err != nil {
		log.Fatal(err)
	}

	err = fileHandle.Close()
	if err != nil {
		log.Fatal(err)
	}

	// Display read information.
	fmt.Println("\nReading IPFS configuration from file: ", fullFileName, "file")
	fmt.Println("Host Name\t: ", configIpfs.HostName)
	fmt.Println("Port\t\t: ", configIpfs.Port)
	fmt.Println("Upload File Path: ", configIpfs.Path)

	return configIpfs
}

// ConnectToIpfs will connect to a IPFS instance,
// based on the read property from an external file.
// It returns a reference to an io.Reader with IPFS instance information.
func ConnectToIpfs(configIpfs ConfigIpfs) *shell.Shell {

	fmt.Println("\nConnecting to IPFS...")

	if configIpfs.HostName == "ipfsHostName" || configIpfs.HostName == "" {
		err1 := errors.New("Invalid HostName")
		log.Fatal("Invalid Hostname error : ", err1)
	}
	// Connect IPFS deamon to IPFS node.
	sh := shell.NewShell(configIpfs.HostName + ":" + configIpfs.Port)

	_, _, errVer := sh.Version()
	if errVer != nil {
		err1 := errors.New("Could not find Daemon running")
		log.Fatal("Daemon error : ", err1)
	}

	// Convert size of chunks into int64.
	givenSize, _ := strconv.ParseInt(configIpfs.ChunkSize, 10, 64)

	if givenSize <= 0 {
		err1 := errors.New("Invalid chunk size entered")
		log.Fatal("Invalid Chunk size : ", err1)
	}

	fmt.Println("Successfully connected to IPFS!")

	return sh
}

// CreateCID will connect to a IPFS instance
// based on the read property from an external file.
// It returns Created CID.
func CreateCID(sh *shell.Shell, data []byte) string {

	readers := bytes.NewReader(data)

	// Create encrypt chunk CID
	encryptChunkCID, err := sh.Add(readers, shell.OnlyHash(true))
	if err != nil {
		log.Fatal("Error in Encryption : ", err)
	}

	// Return IPFS connection object, chunk size and file path.
	return encryptChunkCID
}

// ConnectToIPFSForDownload will connect to a IPFS instance,
// based on the hash name of file on IPFS.
// It returns a reference to an io.Reader with IPFS instance information
func ConnectToIPFSForDownload(hash string, hostName string, port string) *bytes.Reader {

	fmt.Println("\nConnecting to IPFS...")
	if hostName == "ipfsHostName" || hostName == "" {
		err1 := errors.New("Invalid HostName")
		log.Fatal(err1)
	}

	if hash[0:2] != "Qm" || len(hash) != 46 {
		err1 := errors.New("Invalid Shareable Hash")
		log.Fatal(err1)
	}

	// Connect to IPFS daemon to IPFS node.
	sh := shell.NewShell(hostName + ":" + port)
	_, _, errVer := sh.Version()
	if errVer != nil {
		err1 := errors.New("Could not find Daemon running")
		log.Fatal(err1)
	}

	// Inform about successful connection.
	fmt.Println("\nSuccessfully connected to IPFS!")

	// Get data from ipfs node.
	fileReader, err := sh.Cat(hash)
	if err != nil {
		log.Fatal("IPFS data read error: ", err)
	}

	// Read all data recive from ipfs.
	readbytes, _ := ioutil.ReadAll(fileReader)
	reader := bytes.NewReader(readbytes)
	return reader
}

// GetReader returns a Reader of corresponding file whose path is specified.
// io.ReadCloser type of object returned is used to perform transfer of file to Storj.
func GetReader(sh *shell.Shell, configIpfs ConfigIpfs) io.ReadCloser {
	ipfsReader, err1 := os.Open(filepath.Clean(configIpfs.Path))
	if err1 != nil {
		err2 := errors.New("Invalid File path entered")
		log.Fatal("FIle path error : ", err2)
	}
	return ipfsReader
}

// GetReaderDownload returns a Reader of corresponding file whose path is specified.
// io.ReadCloser type of object returned is used to perform transfer of file to Storj.
func GetReaderDownload(sh *shell.Shell, hash string) *bytes.Reader {
	// Get data from ipfs node.
	fileReader, err := sh.Cat(hash)
	if err != nil {
		fmt.Println("IPFS data read error: ", err)
	}

	// // Read all data recive from ipfs.
	readbytes, _ := ioutil.ReadAll(fileReader)
	reader := bytes.NewReader(readbytes)
	return reader
}


cmd/storj.go
package cmd

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"

	"crypto/aes"
	"crypto/cipher"
	"encoding/base64"

	"storj.io/uplink"
)

// ConfigStorj depicts keys to search for within the stroj_config.json file.
type ConfigStorj struct {
	Key                  string `json:"key"`
	APIKey               string `json:"apikey"`
	Satellite            string `json:"satellite"`
	Bucket               string `json:"bucket"`
	UploadPath           string `json:"uploadPath"`
	EncryptionPassphrase string `json:"encryptionpassphrase"`
	SerializedAccess     string `json:"serializedAccess"`
	AllowDownload        string `json:"allowDownload"`
	AllowUpload          string `json:"allowUpload"`
	AllowList            string `json:"allowList"`
	AllowDelete          string `json:"allowDelete"`
	NotBefore            string `json:"notBefore"`
	NotAfter             string `json:"notAfter"`
}

// DownloadConfigStorj structure to store data from json file
type DownloadConfigStorj struct {
	Hash         string `json:"hash"`
	DownloadPath string `json:"downloadPath"`
	Key          string `json:"key"`
}

// LoadStorjConfiguration reads and parses the JSON file that contain Storj configuration information.
func LoadStorjConfiguration(fullFileName string) ConfigStorj {

	var configStorj ConfigStorj
	fileHandle, err := os.Open(filepath.Clean(fullFileName))
	if err != nil {
		log.Fatal("Could not load storj config file: ", err)
	}

	jsonParser := json.NewDecoder(fileHandle)
	if err = jsonParser.Decode(&configStorj); err != nil {
		log.Fatal(err)
	}

	// Close the file handle after reading from it.
	if err = fileHandle.Close(); err != nil {
		log.Fatal(err)
	}

	// Display storj configuration read from file.
	fmt.Println("\nRead Storj configuration from the ", fullFileName, " file")
	fmt.Println("API Key\t\t: ", configStorj.APIKey)
	fmt.Println("Satellite	: ", configStorj.Satellite)
	fmt.Println("Bucket		: ", configStorj.Bucket)

	// Convert the upload path to standard form.
	if configStorj.UploadPath != "" {
		if configStorj.UploadPath == "/" {
			configStorj.UploadPath = ""
		} else {
			checkSlash := configStorj.UploadPath[len(configStorj.UploadPath)-1:]
			if checkSlash != "/" {
				configStorj.UploadPath = configStorj.UploadPath + "/"
			}
		}
	}

	fmt.Println("Upload Path\t: ", configStorj.UploadPath)
	fmt.Println("Serialized Access Key\t: ", configStorj.SerializedAccess)
	return configStorj
}

// LoadStorjDownloadConfiguration reads and parses the JSON file that contain Storj configuration information.
func LoadStorjDownloadConfiguration(fullFileName string) DownloadConfigStorj { // fullFileName for fetching storj V3 credentials from  given JSON filename.

	var downloadConfigStorj DownloadConfigStorj

	fileHandle, err := os.Open(filepath.Clean(fullFileName))
	if err != nil {
		log.Fatal("Error in Opening file")
	}

	jsonParser := json.NewDecoder(fileHandle)
	if err = jsonParser.Decode(&downloadConfigStorj); err != nil {
		log.Fatal(err)
	}

	err = fileHandle.Close()
	if err != nil {
		log.Fatal(err)
	}

	// Display read information.
	fmt.Println("\nReading Download configuration from file: ", fullFileName)
	fmt.Println("Hash\t\t\t: ", downloadConfigStorj.Hash)
	fmt.Println("Download Path\t\t: ", downloadConfigStorj.DownloadPath)

	return downloadConfigStorj
}

// ShareAccess generates and prints the shareable serialized access
// as per the restrictions provided by the user.
func ShareAccess(access *uplink.Access, configStorj ConfigStorj) {

	allowDownload, _ := strconv.ParseBool(configStorj.AllowDownload)
	allowUpload, _ := strconv.ParseBool(configStorj.AllowUpload)
	allowList, _ := strconv.ParseBool(configStorj.AllowList)
	allowDelete, _ := strconv.ParseBool(configStorj.AllowDelete)
	notBefore, _ := time.Parse("2006-01-02_15:04:05", configStorj.NotBefore)
	notAfter, _ := time.Parse("2006-01-02_15:04:05", configStorj.NotAfter)

	permission := uplink.Permission{
		AllowDownload: allowDownload,
		AllowUpload:   allowUpload,
		AllowList:     allowList,
		AllowDelete:   allowDelete,
		NotBefore:     notBefore,
		NotAfter:      notAfter,
	}

	// Create shared access.
	sharedAccess, err := access.Share(permission)
	if err != nil {
		log.Fatal("Could not generate shared access: ", err)
	}

	// Generate restricted serialized access.
	serializedAccess, err := sharedAccess.Serialize()
	if err != nil {
		log.Fatal("Could not serialize shared access: ", err)
	}
	fmt.Println("Shareable serialized access: ", serializedAccess)
}

// ConnectToStorj reads Storj configuration from given file
// and connects to the desired Storj network.
// It then reads data property from an external file.
func ConnectToStorj(fullFileName string, configStorj ConfigStorj, accesskey bool) (*uplink.Access, *uplink.Project) {

	var access *uplink.Access
	var cfg uplink.Config

	// Configure the UserAgent.
	cfg.UserAgent = "IPFS"
	ctx := context.Background()
	var err error

	if accesskey {
		fmt.Println("\nConnecting to Storj network using Serialized access.")
		// Generate access handle using serialized access.
		access, err = uplink.ParseAccess(configStorj.SerializedAccess)
		if err != nil {
			log.Fatal(err)
		}
	} else {
		fmt.Println("\nConnecting to Storj network.")
		// Generate access handle using API key, satellite url and encryption passphrase.
		access, err = cfg.RequestAccessWithPassphrase(ctx, configStorj.Satellite, configStorj.APIKey, configStorj.EncryptionPassphrase)
		if err != nil {
			log.Fatal(err)
		}
	}

	// Open a new porject.
	project, err := cfg.OpenProject(ctx, access)
	if err != nil {
		log.Fatal(err)
	}
	defer project.Close()

	// Ensure the desired Bucket within the Project.
	_, err = project.EnsureBucket(ctx, configStorj.Bucket)
	if err != nil {
		log.Fatal(err)
	}

	fmt.Println("Successfully connected to Storj network.")
	return access, project
}

// UploadData uploads the backup file to storj network.
func UploadData(project *uplink.Project, configStorj ConfigStorj, uploadFileName string, reader io.Reader) {

	ctx := context.Background()

	// Create an upload handle.
	upload, err := project.UploadObject(ctx, configStorj.Bucket, configStorj.UploadPath+uploadFileName, nil)
	if err != nil {
		log.Fatal("Could not initiate upload : ", err)
	}
	fmt.Printf("\nUploading %s to %s.", configStorj.UploadPath+uploadFileName, configStorj.Bucket)

	// Upload data on storj.
	_, err = io.Copy(upload, reader)
	if err != nil {
		abortErr := upload.Abort()
		log.Fatal("Could not upload data to storj: ", err, abortErr)
	}
	// Commit the upload after copying the complete content of the backup file to upload object.
	fmt.Println("\nPlease wait while the upload is being committed to Storj.")
	err = upload.Commit()
	if err != nil {
		log.Fatal("Could not commit object upload : ", err)
	}
}

// DownloadData function downloads the data from storj bucket after upload to verify data is uploaded successfully.
func DownloadData(project *uplink.Project, downloadConfigStorj DownloadConfigStorj, readFile *bytes.Reader) {

	makebuffer := make([]byte, 46)

	// Read data from IPFS
	readHashData, err := readFile.Read(makebuffer)
	if err != nil {
		fmt.Println("Error reading file:", err)
	}

	// Seperate the Hash and configration data
	data := []byte(makebuffer[0:readHashData])
	downloadFileName := string(data)
	downloadFileSize := readFile.Size()

	restDataBuf := make([]byte, downloadFileSize-46)
	readRestFile, err := readFile.Read(restDataBuf)
	if err != nil {
		fmt.Println("Error reading file:", err)
	}

	dataEnc := []byte(restDataBuf[0:readRestFile])
	pkey := []byte(downloadConfigStorj.Key)

	// Decrypt the configration data
	decryptData, err := decrypt(pkey, dataEnc)
	if err != nil {
		log.Fatal(err)
	}

	// Split the configration data
	splitStorjData := strings.Split(string(decryptData), ",")
	downloadBucket := splitStorjData[0]
	downloadPath := splitStorjData[1]
	lastFileName := splitStorjData[2]

	ctx := context.Background()
	var dataDownload []byte
	var lastIndex int64
	var buf = make([]byte, 32768)
	lastIndex = 0

	fmt.Printf("Downloading %s...\n", downloadFileName)

	// Loop to read the object in chunks and store the read data in a byte array.
	download, err3 := project.DownloadObject(ctx, downloadBucket, downloadPath+downloadFileName+"/"+downloadFileName+".txt", &uplink.DownloadOptions{Offset: lastIndex, Length: int64(cap(buf))})
	if err3 != nil {
		fmt.Println("Error: ", err3)
	}

	var err2 error
	dataDownload, err2 = ioutil.ReadAll(download)
	if err2 != nil {
		log.Fatal(err)
	}

	//Convert byte array into String
	receiveContentsMeta := string(dataDownload)

	receiveContentsMeta = strings.TrimSuffix(receiveContentsMeta, ",")

	downloadFileNamesDEBUG := strings.Split(receiveContentsMeta, ",")

	var fileNameDownload = downloadConfigStorj.DownloadPath + "/" + lastFileName

	_, err = os.Stat(fileNameDownload)
	if err == nil {
		err = os.Remove(fileNameDownload)
		if err != nil {
			log.Fatal(err)
		}
	}

	hmkey := []byte("This is a storj ipfs private key")

	for _, filename := range downloadFileNamesDEBUG {
		downloadObj, err := project.DownloadObject(ctx, downloadBucket, downloadPath+downloadFileName+"/"+filename, &uplink.DownloadOptions{Offset: lastIndex, Length: int64(cap(buf))})
		if err != nil {
			log.Fatalf("Could not open object at %q: %v", downloadPath+downloadFileName+"/"+filename, err)
		}
		if _, err = os.Stat("./debug"); os.IsNotExist(err) {
			err1 := os.Mkdir("./debug", 0750)
			if err1 != nil {
				log.Fatal("Could not create debug folder: ", err1)
			}
		}

		// Read everything from the stream.
		receivedContents, err := ioutil.ReadAll(downloadObj)
		if err != nil {
			log.Fatal("Could not Read All content in stream:", err)
		}

		//Decryt the downloaded file data from storj
		dec, err := decrypt(hmkey, receivedContents)
		if err != nil {
			log.Fatal("Could not decrypt received data:", err)
		}
		// Create/open file in append mode.
		downloadFileDisk, err := os.OpenFile(filepath.Clean(filepath.Join("./debug", lastFileName)), os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0600)
		if err != nil {
			log.Fatal(err)
		}

		// read data
		dataReader := bytes.NewReader(dec)
		_, err = io.Copy(downloadFileDisk, dataReader)
		if err != nil {
			log.Fatal(err)
		}

		// Close the file handle after reading from it.
		if err = downloadFileDisk.Close(); err != nil {
			log.Fatal(err)
		}
	}
	fmt.Printf("File downloading: Complete!\n")
	fmt.Printf("\n file \"%s\" downloaded to \"%s\"\n", lastFileName, "debug/")
}

// Function to decrypt data based on given key.
func decrypt(key, text []byte) ([]byte, error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}
	if len(text) < aes.BlockSize {
		return nil, errors.New("ciphertext too short")
	}
	iv := text[:aes.BlockSize]
	text = text[aes.BlockSize:]
	cfb := cipher.NewCFBDecrypter(block, iv)
	cfb.XORKeyStream(text, text)
	data, err := base64.StdEncoding.DecodeString(string(text))
	if err != nil {
		return nil, err
	}
	//nolint:ineffassign
	iv = nil
	text = nil
	cfb = nil
	return data, nil
}


cmd/store.go
package cmd

import (
	"bytes"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"os"
	"path/filepath"
	"strconv"

	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"encoding/base64"

	shell "github.com/ipfs/go-ipfs-api"
	chunker "github.com/ipfs/go-ipfs-chunker"
	"github.com/spf13/cobra"
)

// storeCmd represents the store command.
var storeCmd = &cobra.Command{
	Use:   "store",
	Short: "Command to upload data to storj V3 network.",
	Long:  `Command to connect to desired IPFS account and back-up the complete data to given Storj Bucket.`,
	Run:   ipfsStore,
}

// DownCmd represents the download command.
var DownCmd = &cobra.Command{
	Use:   "download",
	Short: "Command to download data from storj V3 network.",
	Long:  `Command to download data from the Storj Bucket using the Hash.`,
	Run:   storjDownload,
}

func init() {

	// Setup the store command with its flags.
	rootCmd.AddCommand(storeCmd)
	rootCmd.AddCommand(DownCmd)
	var defaultIpfsFile string
	var defaultStorjFile string
	var defaultStorjDownloadFile string
	storeCmd.Flags().BoolP("accesskey", "a", false, "Connect to storj using access key(default connection method is by using API Key).")
	storeCmd.Flags().BoolP("share", "s", false, "For generating share access of the uploaded backup file.")
	storeCmd.Flags().StringVarP(&defaultIpfsFile, "ipfs", "i", "././config/ipfs_property.json", "full filepath contaning IPFS configuration.")
	storeCmd.Flags().StringVarP(&defaultStorjFile, "storj", "u", "././config/storj_config.json", "full filepath contaning storj V3 configuration.")
	DownCmd.Flags().StringVarP(&defaultIpfsFile, "ipfs", "i", "././config/ipfs_property.json", "full filepath contaning IPFS configuration.")
	DownCmd.Flags().BoolP("accesskey", "a", false, "Connect to storj using access key(default connection method is by using API Key).")
	DownCmd.Flags().StringVarP(&defaultStorjFile, "storj", "u", "././config/storj_config.json", "full filepath contaning storj V3 configuration.")
	DownCmd.Flags().StringVarP(&defaultStorjDownloadFile, "storjDown", "d", "././config/storj_download.json", "Download data from stroj")
}

func ipfsStore(cmd *cobra.Command, args []string) {

	// Process arguments from the CLI.
	ipfsConfigfilePath, _ := cmd.Flags().GetString("ipfs")
	fullFileNameStorj, _ := cmd.Flags().GetString("storj")
	useAccessKey, _ := cmd.Flags().GetBool("accesskey")
	useAccessShare, _ := cmd.Flags().GetBool("share")

	// Read IPFS instance's configurations from an external file and create an IPFS configuration object.
	configIpfs := LoadIpfsProperty(ipfsConfigfilePath)

	// Read storj network configurations from and external file and create a storj configuration object.
	storjConfig := LoadStorjConfiguration(fullFileNameStorj)

	// Connect to storj network using the specified credentials.
	access, project := ConnectToStorj(fullFileNameStorj, storjConfig, useAccessKey)

	// Connect to IPFS using the specified credentials
	ipfsShell := ConnectToIpfs(configIpfs)

	fileHandle := GetReader(ipfsShell, configIpfs)

	fmt.Println("\nReading content from the file:", configIpfs.Path)

	// Get file name from the file path from configration file.
	_, lastFileName := filepath.Split(configIpfs.Path)

	// Create encrypt Base CID
	encryptCID, _ := ipfsShell.Add(fileHandle, shell.OnlyHash(true))

	// Open the uploaded file
	file, err1 := os.Open(configIpfs.Path)
	if err1 != nil {
		fmt.Println(err1)
	}

	// Get total size of uploaded file
	statFile, err4 := file.Stat()
	if err4 != nil {
		fmt.Println(err4)
	}
	fileSize := statFile.Size()

	givenSize, _ := strconv.ParseInt(configIpfs.ChunkSize, 0, 64)

	// Generate the number of chunk files
	noOfChunkFiles := int(fileSize/givenSize) + 1

	// Divided total uploaded file data into chunks DAG.
	chunkFile := chunker.NewSizeSplitter(file, givenSize)

	var metaFile *os.File
	metaFileName := "./metadata.txt"

	newFile, err2 := os.Create(metaFileName)
	if err2 != nil {
		log.Fatal(err2)
	}

	err2 = newFile.Close()
	if err2 != nil {
		log.Fatal(err2)
	}

	for i := 0; i < noOfChunkFiles; i++ {

		// Get the chunks data from the chunks DAG.
		storeChunkFile, _ := chunkFile.NextBytes()

		//Encrypt the chunk data by the given key
		key := []byte("This is a storj ipfs private key")

		encryptData, err := encrypt(key, storeChunkFile)
		if err != nil {
			log.Fatal(err)
		}

		// Create chunk CID using bytes data
		encryptChunkCID := CreateCID(ipfsShell, encryptData)

		fileName := encryptCID + "/" + encryptChunkCID

		reader := bytes.NewReader(encryptData)

		// Upload chunk data on storj Network with baseCID/chunkCID name.
		UploadData(project, storjConfig, fileName, reader)

		// Write all chunks CID into loacl disk file in append mode.
		metaFile, _ = os.OpenFile(metaFileName, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0600)
		if _, err := metaFile.WriteString(encryptChunkCID + ","); err != nil {
			log.Fatal(err)
		}

		// Close meta file after write all chunks into the file
		err = metaFile.Close()
		if err != nil {
			log.Fatal(err)
		}
	}

	// Open meta file from local disk
	openMetaFile, _ := os.Open(metaFileName)
	metaBytes, err := ioutil.ReadAll(openMetaFile)
	if err != nil {
		log.Fatal("Reader Error:", err)
	}
	metaReader := bytes.NewReader(metaBytes)
	// Close Meta file
	err = openMetaFile.Close()
	if err != nil {
		log.Fatal(err)
	}
	// Remove meta file from local disk.
	err = os.Remove(metaFileName)
	if err != nil {
		log.Fatal(err)
	}

	metaFileStoreName := encryptCID + "/" + encryptCID + ".txt"

	// Store meta file data on storj network with baseCID/baseCID.txt
	UploadData(project, storjConfig, metaFileStoreName, metaReader)

	fmt.Println("\nAdding configuration data to IPFS: Initiated...")

	ipfsStorjData := storjConfig.Bucket + "," + storjConfig.UploadPath + "," + lastFileName

	//Encrypt the storj configration data
	enkey := []byte(storjConfig.Key)

	ipfsStorjDataBytes := []byte(ipfsStorjData)
	storjEncryptData, err := encrypt(enkey, ipfsStorjDataBytes)
	if err != nil {
		log.Fatal(err)
	}

	hash := []byte(encryptCID)

	// Create buffer for Chunk CID and encrypted Storj configurations.
	encryptedStorjConfig := append(hash, storjEncryptData...)

	// Create the CID from encrypted chunk data and encrypted
	// storj configration and enrypted private key.
	configHash, _ := ipfsShell.Add(bytes.NewReader(encryptedStorjConfig))
	fmt.Println("Shareable Hash:", configHash)

	// Create restricted shareable serialized access if share is provided as argument.
	if useAccessShare {
		ShareAccess(access, storjConfig)
	}
}

func storjDownload(cmd *cobra.Command, args []string) {

	// Process arguments from the CLI.
	ipfsConfigfilePath, _ := cmd.Flags().GetString("ipfs")
	fullFileNameDownload, _ := cmd.Flags().GetString("storjDown")
	useAccessKey, _ := cmd.Flags().GetBool("accesskey")
	fullFileNameStorj, _ := cmd.Flags().GetString("storj")

	// Read storj network configurations from and external file and create a storj configuration object.
	storjConfig := LoadStorjConfiguration(fullFileNameStorj)

	// Read IPFS instance's configurations from an external file and create an IPFS configuration object.
	configIpfs := LoadIpfsProperty(ipfsConfigfilePath)

	// Connect to storj network using the specified credentials.
	_, project := ConnectToStorj(fullFileNameStorj, storjConfig, useAccessKey)

	// Read storj network cofiguration related to download.
	downloadConfig := LoadStorjDownloadConfiguration(fullFileNameDownload)

	// Connect to ipfs network using specified credentials.
	ipfsDownloadShell := ConnectToIpfs(configIpfs)

	reader := GetReaderDownload(ipfsDownloadShell, downloadConfig.Hash)

	DownloadData(project, downloadConfig, reader)

}

// Encrypt Function to encrypt data with specified key
func encrypt(key, text []byte) ([]byte, error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}
	b := base64.StdEncoding.EncodeToString(text)
	ciphertext := make([]byte, aes.BlockSize+len(b))
	iv := ciphertext[:aes.BlockSize]
	if _, err := io.ReadFull(rand.Reader, iv); err != nil {
		return nil, err
	}
	cfb := cipher.NewCFBEncrypter(block, iv)
	cfb.XORKeyStream(ciphertext[aes.BlockSize:], []byte(b))
	return ciphertext, nil
}
